heat/core/linalg/basics.py

```diff
@@ -1075,6 +1077,317 @@ def projection(a, b):
+
+            # empty diagonal => create an array of zeros for following summation
+            if 0 in diag_t.shape:
+                res_shape = [1 if i == 0 else i for i in diag_t.shape]
+                diag_t = torch.zeros(res_shape, device=a.device.torch_device)
+
+            # create recvbuffer (with correct resulting shape)
+            sum_along_diagonals_t = torch.clone(diag_t)
+            res_shape = list(sum_along_diagonals_t.shape)
+            del res_shape[-1]  # as summed up along the last axis
+            sum_along_diagonals_t = torch.reshape(sum_along_diagonals_t, res_shape)
+
+            # Sum up all partial sums (and gather them)
+            a.comm.Allreduce(MPI.IN_PLACE, sum_along_diagonals_t, MPI.SUM)
+
+            # Store result in out if provided
+            if out is not None:
+                if out.split is not None:
+                    warnings.warn(
+                        f"Split axis of `out` will be changed from {out.split} to None to "
+                        f"guarantee correct results."
+                    )
+                    # Allgather of out - necessary as whole result is on all processes after Allreduce
+                    out.resplit_(None)
+                sanitation.sanitize_out(out, tuple(res_shape), out.split, a.device)
+                out.larray = sum_along_diagonals_t
+                return out
+
+            last_axis = sum_along_diagonals_t.ndim - 1
+            split_axis = a.split if a.split <= last_axis else last_axis
+
+            sum_along_diagonals = factories.array(
+                sum_along_diagonals_t, dtype=dtype, split=split_axis, comm=a.comm, device=a.device
+            )
+
+            return sum_along_diagonals
+
+        if a.is_distributed():
+            # (...and a.split not in (axis1, axis2))
+            if a.split < axis2:
+                gather_axis = a.split
+            else:
+                gather_axis = a.split - 2
+
+            # check if gather_axis is in range of result
+            if gather_axis >= sum_along_diagonals_t.ndim:
+                gather_axis = sum_along_diagonals_t.ndim - 1
+
+            # Stack all partial results back together along the correct axis
+            sum_along_diagonals = factories.array(
+                sum_along_diagonals_t,
+                dtype=dtype,
+                is_split=gather_axis,
+                comm=a.comm,
+                device=a.device,
+            )
+        # input not distributed
+        else:
+            # check if split axis is in range of result
+            if a.split is not None and a.split >= sum_along_diagonals_t.ndim:
+                gather_axis = sum_along_diagonals_t.ndim - 1
+            else:
+                gather_axis = a.split
+
+            # convert torch result back to DNDarray
+            sum_along_diagonals = factories.array(
+                sum_along_diagonals_t, dtype=dtype, split=gather_axis, comm=a.comm, device=a.device
+            )
```

Comment:
to clarify, you will return a split object even if the input is `split=None`?

Reply:
No, this won't happen. If `a.split=None`, the first part of the (`and`) if condition is already false and the else case is always executed. Thus the result will also be not distributed.